File: medium/2770-maximum-number-of-jumps-to-reach-the-last-index/solution.py
from typing import List


class Solution:
    def maximumJumps(self, nums: List[int], target: int) -> int:
        """
        2770. Maximum Number of Jumps to Reach the Last Index
        Time: O(n²)
        Space: O(n)
        """
        n = len(nums)
        dp = [-1] * n
        dp[0] = 0

        for i in range(n):
            if dp[i] == -1:
                continue

            for j in range(i + 1, n):
                if abs(nums[j] - nums[i]) <= target:
                    dp[j] = max(dp[j], dp[i] + 1)

        return dp[-1]

File: medium/2770-maximum-number-of-jumps-to-reach-the-last-index/solution.java
/**
 * 2770. Maximum Number of Jumps to Reach the Last Index
 * Time: O(n²)
 * Space: O(n)
 */
class Solution {
    public int maximumJumps(int[] nums, int target) {
        int n = nums.length;
        int[] dp = new int[n];

        for (int i = 1; i < n; i++) {
            dp[i] = -1;
        }

        for (int i = 0; i < n; i++) {
            if (dp[i] == -1) {
                continue;
            }

            for (int j = i + 1; j < n; j++) {
                long diff = (long) nums[j] - nums[i];
                if (Math.abs(diff) <= target) {
                    dp[j] = Math.max(dp[j], dp[i] + 1);
                }
            }
        }

        return dp[n - 1];
    }
}
